### ISSUE TITLE
Docker ENV variables - how to properly set them? 

### ISSUE THREAD

**Issue by stlbucket** [nuxt-modules/supabase#269]
I think this is the same everywhere in source:
<img width="1051" alt="image" src="https://github.com/nuxt-modules/supabase/assets/2467892/18f54a82-ec54-4118-92ba-311c2ec44182">

how do we get these values into runtime config?

The documentation does not refer to any runtime config fields - rather it use .env file and also nuxt.config.supabase.url and key.

But how do we set these in a docker environment?  How do we specify runtime ENV variables so they take proper effect?

I am assuming it merges somehow, but I cannot figure it out.

<img width="745" alt="image" src="https://github.com/nuxt-modules/supabase/assets/2467892/83183090-25d8-4ae5-87cc-13749b862190">

**Comment by stlbucket** [comment]
I have written an api route to try and mimic the nuxt/supabase code:

```
export default defineEventHandler((event) => {
  const rtc = useRuntimeConfig()
  console.log(rtc)
  const {
    supabase: { url, key, cookieName },
  } = useRuntimeConfig().public
  return {
    ...rtc,
    supabase: {
      url: url,
      key: key,
      cookieName: cookieName
    }
  }
})
```

an error is thrown when trying to destructure out of runtimeConfig

```
{
  app: { baseURL: '/', buildAssetsDir: '/_nuxt/', cdnURL: '' },
  nitro: { envPrefix: 'NUXT_', routeRules: { '/__nuxt_error': [Object] } },
  public: {
    SUPABASE_URL: 'http://localhost:54321',
    SUPABASE_KEY: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6ImFub24iLCJleHAiOjE5ODM4MTI5OTZ9.CRXP1A7WOeoJeXxjNni43kdQwgnWNReilDMblYTn_I0',
    'graphql-client': { codegen: false },
    GQL_HOST: 'http://localhost:3025/api/graphql'
  },
  SUPABASE_URI: 'postgresql://postgres:postgres@localhost:54322/postgres',
  SUPABASE_SERVICE_KEY: 'eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZS1kZW1vIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImV4cCI6MTk4MzgxMjk5Nn0.EGIM96RAZx35lJzdJsyH-qQwv8Hdp7fsn3W0YpN81IU',
  SUPABASE_JWT_SECRET: 'super-secret-jwt-token-with-at-least-32-characters-long'
}
[nuxt] [request error] [unhandled] [500] Cannot read properties of undefined (reading 'url')
  at <anonymous> (./server/api/rtc.ts:5:1)
  at ./node_modules/h3/dist/index.mjs:1563:14
  at Object.handler (./node_modules/h3/dist/index.mjs:1806:28)
  at Object.handler (./node_modules/h3/dist/index.mjs:1630:31)
  at async toNodeHandle (./node_modules/h3/dist/index.mjs:1840:7)
  at async ufetch (./node_modules/unenv/runtime/fetch/index.mjs:9:17)
  at async $fetchRaw2 (./node_modules/ofetch/dist/shared/ofetch.00501375.mjs:219:26)
  at async $fetchRaw2 (./node_modules/ofetch/dist/shared/ofetch.00501375.mjs:256:14)
  at async $fetch2 (./node_modules/ofetch/dist/shared/ofetch.00501375.mjs:261:15)
[nitro] [unhandledRejection] ReferenceError: alert is not defined
    at loadData (/Users/buckfactor/function-bucket/env-var/app.vue:21:9)
```

**Comment by stlbucket** [comment]
I see how this should be working in the module.ts now.  But still it does not work without a .env file.

**Comment by stlbucket** [comment]
<img width="628" alt="image" src="https://github.com/nuxt-modules/supabase/assets/2467892/d558182a-8c87-4be3-8f87-737676ce8c1d">

I am assuming this is happening because useRuntimeConfig (composables) isn't available in modules.  But is this use of process.env correct or is it just short-circuiting for local dev?

If useRuntimeConfig won't work, will these process.env variables be available in all environments?  They do not appear to be available in my docker build.

**Comment by stlbucket** [comment]
In my example, enabling the nuxt module causes the site to error saying SupabaseUrl is required:

<img width="675" alt="image" src="https://github.com/nuxt-modules/supabase/assets/2467892/f083d7a6-c929-40a7-8ca8-251e43702255">

and in /server/api/rtc.ts:

<img width="501" alt="image" src="https://github.com/nuxt-modules/supabase/assets/2467892/a20dcb8b-4991-4d09-a85c-52e0c3640ea5">

**Comment by stlbucket** [comment]
this usage of process.env.SUPABASE_XXX seems like maybe a bad idea just for this?
```
    const { url, key, cookieName, clientOptions} = useRuntimeConfig().public.supabase
```
this code isn't that much more cumbersome, I think.
```
    const url = useRuntimeConfig().public.SUPABASE_URL
    const key = useRuntimeConfig().public.SUPABASE_KEY
    const serviceKey = useRuntimeConfig().SUPABASE_SERVICE_KEY
```
or
```
  const { SUPABASE_URL, SUPABASE_KEY } = useRuntimeConfig().public
  const { SUPABASE_SERVICE_KEY } = useRuntimeConfig()
```
are there any other side effects of this construct?

is there something fundamental i am missing here?

**Comment by stlbucket** [comment]
also - from my cloud run service yaml:
<img width="208" alt="image" src="https://github.com/nuxt-modules/supabase/assets/2467892/82aea837-e4d0-44c8-b9f2-819527f2fff7">

just the NUXT_ prefixes are necessary and they are passed along just fine according to nuxt.config runtime rules

**Comment by stlbucket** [comment]
I have also tried providing redundant SUPABASE_KEY and SUPABASE_URL env variables alongside the NUXT_*** values with no luck

**Comment by stlbucket** [comment]
https://github.com/stlbucket/nuxt-supabase-stlbucket

I've created a repository with a possible reworking of the instantiation.  Everything works in the playground, but currently trying to incorporate into my project locally.

**Comment by stlbucket** [comment]
ugh - pretty sure this is an env naming issue

I was not prefixing with NUXT_

**Comment by stlbucket** [comment]
TLDR:

In docker, you may need two settings:

NUXT_PUBLIC_SUPABASE_URL - so it shows up in runtime config
SUPABASE_URL - the module needs this on startup

**Comment by NLDylan** [comment]
@stlbucket I see this issue is closed, but not really sure what the fix is.
Currently i am also deploying a nuxt 3 project with docker on digital ocean.
All the other environment variables are working expect for the nuxt/supabase module.

What did you do to fix this?
Added ARG or ENV options to the docker build process or docker file?

**Comment by stlbucket** [comment]
I just added both ENV variables:

NUXT_PUBLIC_SUPABASE_URL - so it shows up in runtime config
SUPABASE_URL - the module needs this on startup

In trying to figure this out I looked into the source for nuxt/supabase, and it uses the naked variable SUPABASE_URL during initialization and then later uses useRuntimeConfig(), which comes from NUXT_PUBLIC_SUPABASE_URL to instantiate the client.

in the repository linked above, which is forked from this one, i played around with trying to make this more consistent.  i'm not sure if that's a useful concept or not.  it seems to work with the playground, and would cause less confusion as it would be more consistent with nuxt docs, in general.  but there are surely some reasons i am unaware of as to why it is the way it is.

**Comment by L422Y** [comment]
It's crazy that the documentation is so lacking with this, so close to being perfect, all we need is better documentation on ENVs, or better/more verbose errors.

**Comment by stlbucket** [comment]
@NLDylan - on digital ocean, i just set ENV variables

**Comment by WebDeveloperBen** [comment]
I also had a similar issue where i had to set both SUPABASE_SERVICE_KEY and NUXT_SUPABASE_SERVICE_KEY to have it available inside my docker container. Didn't have that issue with vercel mind you.

**Comment by destel** [comment]
I spent an evening trying to figure this out. I couldn't make it work with NUXT_* variables alone. And with unprefixed variables there are two issues:
1. All values including service key are baked into the server bundle (can be confirmed by searching in .output dir)
2. On cloudlare workers I had to use build time variables/secrets instead of "normal" ones

Is this an intended behavior?
